File: JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/dgraph/CustomDgraphExpandedInteraction.java
package org.jembi.jempi.libmpi.dgraph;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.jembi.jempi.shared.models.CustomDemographicData;
import org.jembi.jempi.shared.models.ExpandedInteraction;
import org.jembi.jempi.shared.models.Interaction;

import java.util.List;

@JsonInclude(JsonInclude.Include.NON_NULL)
record CustomDgraphExpandedInteraction(
      @JsonProperty("uid") String patientId,
      @JsonProperty("Interaction.source_id") DgraphSourceId sourceId,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_AUX_ID) String auxId,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_GIVEN_NAME) String givenName,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_FAMILY_NAME) String familyName,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_GENDER) String gender,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_DOB) String dob,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_CITY) String city,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_PHONE_NUMBER) String phoneNumber,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_NATIONAL_ID) String nationalId,
      @JsonProperty("~GoldenRecord.interactions") List<CustomDgraphReverseGoldenRecord> dgraphGoldenRecordList) {

   Interaction toInteraction() {
      return new Interaction(this.patientId(),
                             this.sourceId().toSourceId(),
                             new CustomDemographicData(
                                     this.auxId(),
                                     this.givenName(),
                                     this.familyName(),
                                     this.gender(),
                                     this.dob(),
                                     this.city(),
                                     this.phoneNumber(),
                                     this.nationalId()));
   }

   ExpandedInteraction toExpandedInteraction() {
      return new ExpandedInteraction(this.toInteraction(),
                                     this.dgraphGoldenRecordList()
                                         .stream()
                                         .map(CustomDgraphReverseGoldenRecord::toGoldenRecordWithScore)
                                         .toList());
   }

}

File: JeMPI_Apps/JeMPI_Configuration/src/main/scala/configuration/CustomDgraphExpandedInteraction.scala
package configuration

import java.io.{File, PrintWriter}

private object CustomDgraphExpandedInteraction {

  private val classLocation = "../JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/dgraph"
  private val customClassName = "CustomDgraphExpandedInteraction"
  private val packageText = "org.jembi.jempi.libmpi.dgraph"

  def generate(config: Config): Unit =
    val classFile: String = classLocation + File.separator + customClassName + ".java"
    println("Creating " + classFile)
    val file: File = new File(classFile)
    val writer: PrintWriter = new PrintWriter(file)
    writer.println(
      s"""package $packageText;
         |
         |import com.fasterxml.jackson.annotation.JsonInclude;
         |import com.fasterxml.jackson.annotation.JsonProperty;
         |import org.jembi.jempi.shared.models.CustomUniqueInteractionData;
         |import org.jembi.jempi.shared.models.CustomDemographicData;
         |import org.jembi.jempi.shared.models.ExpandedInteraction;
         |import org.jembi.jempi.shared.models.Interaction;
         |
         |import java.util.List;
         |
         |@JsonInclude(JsonInclude.Include.NON_NULL)
         |record $customClassName(
         |      @JsonProperty("uid") String interactionId,
         |      @JsonProperty("Interaction.source_id") DgraphSourceId sourceId,
         |${interactionFields()}
         |      @JsonProperty("~GoldenRecord.interactions") List<CustomDgraphReverseGoldenRecord> dgraphGoldenRecordList) {
         |
         |   Interaction toInteraction() {
         |      return new Interaction(this.interactionId(),
         |                             this.sourceId().toSourceId(),
         |                             new CustomUniqueInteractionData(${uniqueArguments()}),
         |                             new CustomDemographicData(${demographicArguments()}));
         |   }
         |
         |   ExpandedInteraction toExpandedInteraction() {
         |      return new ExpandedInteraction(this.toInteraction(),
         |                                     this.dgraphGoldenRecordList()
         |                                         .stream()
         |                                         .map(CustomDgraphReverseGoldenRecord::toGoldenRecordWithScore)
         |                                         .toList());
         |   }
         |
         |}
         |""".stripMargin)
    writer.flush()
    writer.close()


    def interactionFields(): String =

      def mapField(fieldName: String, fieldType: String): String = s"""      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_${fieldName.toUpperCase}) ${Utils.javaType(fieldType)} ${Utils.snakeCaseToCamelCase(fieldName)},"""

      val f1 = if (config.uniqueInteractionFields.isEmpty) "" else
        config
          .uniqueInteractionFields
          .get
          .map(f => mapField(f.fieldName, f.fieldType))
          .mkString("\n") + "\n"

      val f2 = config
        .demographicFields
        .map(f => mapField(f.fieldName, f.fieldType))
        .mkString("\n")

      f1 + f2
    end interactionFields

    def uniqueArguments(): String =
      if (config.uniqueInteractionFields.isEmpty)
        ""
      else
        config
          .uniqueInteractionFields
          .get
          .map(f =>
            s"""${" " * 63}this.${Utils.snakeCaseToCamelCase(f.fieldName)}(),""")
          .mkString("\n").trim.dropRight(1)
    end uniqueArguments

    def demographicArguments(): String =
      config
        .demographicFields
        .map(f =>
          s"""${" " * 55}this.${Utils.snakeCaseToCamelCase(f.fieldName)}(),""")
        .mkString("\n").trim.dropRight(1)
    end demographicArguments

  end generate

}
